### PR TITLE
Build dependency resolution with target_kind filter

### PR DESCRIPTION
**Pull request by maciektr**
commit-id:ade97ea3

---

**Stack**:
- #679
- #651
- #649
- #628
- #660
- #615
- #597
- #659
- #612 ⬅

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

File: scarb/src/core/resolver.rs [hunk 1]
```diff
@@ -38,4 +39,57 @@ impl Resolve {
             .iter(&self.graph)
             .unique()
     }
+
+    pub fn solution_with_target_kind(
+        &self,
+        root_package: PackageId,
+        target_kind: TargetKind,
+    ) -> Vec<PackageId> {
```
Comment (maciektr) [inline scarb/src/core/resolver.rs:47]: Maybe we could optimise it to not collect vec? 🤔

Comment (mkaput) [reply scarb/src/core/resolver.rs:47]: why not?

File: scarb/src/core/resolver.rs [hunk 2]
```diff
@@ -38,4 +39,57 @@ impl Resolve {
             .iter(&self.graph)
             .unique()
     }
+
+    pub fn solution_with_target_kind(
```
Comment (mkaput) [inline scarb/src/core/resolver.rs:43]: Are we calling `solution_of` anywhere now?

File: scarb/src/core/resolver.rs [hunk 3]
```diff
@@ -38,4 +39,57 @@ impl Resolve {
...
+
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct DependencyEdge(SmallVec<[TargetKind; 4]>);
+
+impl DependencyEdge {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn accepts_target_kind(&self, target_kind: TargetKind) -> bool {
```
Comment (mkaput) [inline scarb/src/core/resolver.rs:67]:
```suggestion
    pub fn accepts_target(&self, target_kind: TargetKind) -> bool {
```

IMHO pushing `_kind` everywhere is superfluous

File: scarb/src/core/resolver.rs [hunk 4]
```diff
@@ -38,4 +39,57 @@ impl Resolve {
...
+        });
+        Dfs::new(&filtered_graph, root_package)
+            .iter(&filtered_graph)
+            .unique()
+            .collect_vec()
+    }
+}
+
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct DependencyEdge(SmallVec<[TargetKind; 4]>);
```
Comment (mkaput) [inline scarb/src/core/resolver.rs:60]: Why 4 instead of 1? Can't we represent wildcards here?

**Review by mkaput — CHANGES_REQUESTED** [review]

File: scarb/src/core/resolver.rs (hunk 4 above)
Comment (maciektr) [reply scarb/src/core/resolver.rs:60]: Right now empty vec = wildcard, hence https://github.com/software-mansion/scarb/blob/ae88e56894898b59005a6f7de1361c1829df07a3/scarb/src/core/resolver.rs#L79

File: scarb/src/core/resolver.rs (hunk 1 above)
Comment (maciektr) [reply scarb/src/core/resolver.rs:47]: The iterator references values owned by `filtered_graph` variable, I haven't wrapped my head how to reorganize the code yet

Comment (mkaput) [reply scarb/src/core/resolver.rs:47]: In general, the solutions here won't be massive, and it's not really that important to try to optimize out this vec collection. If that's hard, then let's skip trying to do that now.

File: scarb/src/core/package/name.rs
Comment (mkaput) [inline scarb/src/core/package/name.rs:1]: not in scope of this PR?

File: scarb/src/ops/compile.rs [hunk 5]
```diff
@@ -32,7 +32,7 @@ pub fn compile(packages: Vec<PackageId>, ws: &Workspace<'_>) -> Result<()> {
 }
 
 fn compile_unit(unit: CompilationUnit, ws: &Workspace<'_>) -> Result<()> {
-    let package_name = unit.main_package_id.name.clone();
+    let package_name = unit.main_package_id.name.to_string();
```
Comment (mkaput) [inline scarb/src/ops/compile.rs:35]: why?

**Review by mkaput — APPROVED** [review]

**Review by mkaput — APPROVED** [review]

**Pull request merged**